### PR TITLE
feat(image-map-action-anim): export/import parity (closes #499)

### LINKED ISSUE #499: feat(image-map-action-anim): Avalonia export/import parity
## Summary

Add cross-platform export/import for Map Action Animations to the Avalonia GUI.

WinForms `ImageUtilMapActionAnimation.Export/ExportGif/Import` produce `.MapActionAnimation.txt` scripts + animated GIFs, and consume the same files for import. These live in the WinForms project, depend on `System.Drawing.Bitmap` and `OpenFileDialog`/`SaveFileDialog`, and have no Core equivalent.

`FEBuilderGBA.Core/ImageUtilMapActionAnimationCore.cs` currently has only `DrawFrame/CountFrames` — it cannot encode/decode the script format or build GIFs.

## Scope

- Extract `ImageUtilMapActionAnimation.Export` (script format) into Core, using `IImage` + `CoreState.ImageService` instead of `System.Drawing.Bitmap`.
- Extract `ImageUtilMapActionAnimation.ExportGif` into Core or `FEBuilderGBA.SkiaSharp` (SkiaSharp can encode GIFs).
- Extract `ImageUtilMapActionAnimation.Import` into Core (parses script, decompresses tile/palette data, writes ROM).
- Wire Avalonia view buttons: `AnimationExportButton`, `AnimationImportButton`, `OpenSourceButton`, `SelectSourceButton`, with `StorageProvider` (`SaveFilePicker`/`OpenFilePicker`) instead of WinForms dialogs.
- Track imported source-file path in `Program.ResourceCache` for the Open Source / Select Source buttons.

## Background

Filed during gap-sweep fix for #433 — see https://github.com/laqieer/FEBuilderGBA/pull/<N> Known Limitations section. The 4 WF-only labels are deferred to this issue: アニメーション取出 / アニメーション読込 / ソースファイルを開く / ソースフォルダーを開く.

Refs #374, #433.

---

Generated with [Claude Code](https://claude.com/claude-code) (claude-opus-4-7[1m]).

### PR DESCRIPTION
**Pull request by laqieer**
## Summary

- Add `MapActionAnimationExportImportCore` to Core that ports the WinForms `ImageUtilMapActionAnimation.Export` / `ExportGif` / `Import` trio cross-platform. Wires the four WF-only labels deferred from #506 — `アニメーション取出` / `アニメーション読込` / `ソースファイルを開く` / `ソースフォルダーを開く` — as functional Avalonia buttons.
- Move `GifEncoder` from `FEBuilderGBA.CLI` to `FEBuilderGBA.Core` as `GifEncoderCore`. Adds `IndexedToRgba(indexedPixels, paletteRgba, w, h)` so callers that feed indexed `IImage` (e.g. `Decode4bppTiles` output) can expand to RGBA before encoding. Adds Core `U.GameFrameSecToGifFrameSec(uint)` that mirrors the WinForms `Math.Round` semantics (wait=1 -> 2cs).
- Avalonia view: 4 new buttons with proper `AutomationId`s. Export gated on `IsAnimationValid`; Import gated on `IsLoaded && !IsEmptyEntry`; OpenSource/SelectSource gated on remembered source path (via `CoreState.ResourceCache` keyed `MapActionAnimation_<hex-id>`). Import wraps ROM mutations in `UndoService.Begin/Commit/Rollback`.

## Plan Reference

Implements the v3 plan from https://github.com/laqieer/FEBuilderGBA/issues/499#issuecomment-4529340461 (Copilot CLI plan-review pass 3 accepted: https://github.com/laqieer/FEBuilderGBA/issues/499#issuecomment-4529400000 — no remaining blockers from passes 1+2).

All 5 + 2 + 0 = 7 plan-review concerns addressed:
- **C1/B1** (RGBA conversion): `DecreaseColorCore.Quantize` + `GifEncoderCore.IndexedToRgba` for indexed→RGBA.
- **C2** (palette preservation): inline `# RAW-TILES:` / `# RAW-PALETTE:` comments in the .txt give byte-equal round-trip.
- **C3** (free-space placement): `ImageImportCore.FindAndWriteData` (not raw `AppendToRomEnd`).
- **C4** (GIF delay rounding): `U.GameFrameSecToGifFrameSec` (Math.Round) replaces truncating division.
- **C5** (dedup collision): `Dictionary<(uint imgPtr, uint palPtr), string>` tuple key, not arithmetic sum.
- **B2** (palette padding): `PadGBAPaletteTo16` guarantees exactly 32 bytes for any quantized palette.
- **A3** (RAW comment parsing order): RAW-* checked BEFORE generic `U.IsComment` skip; each frame line carries its own preceding RAW block for duplicate-reference round-trip.

## Scope

Closes #499.

Ref #374, #433 (#499 is the last of the three follow-ups filed during PR #506 — #500 / #501 remain open).

## Screenshots

Real Avalonia GUI capture via `PrintWindow` API + `UIAutomationClient` against `roms/FE8U.gba`. The screenshot is commit-pinned on master via #618 so the URL is permanent.

ImageMapActionAnimation editor — Export/Import/Source-file buttons wired (#499)

The capture demonstrates the new control surface:
- **Export Animation** button (rendered, disabled — `IsAnimationValid=false` on vanilla FE8U with no map-action-animation patch installed).
- **Import Animation** button (rendered, disabled — `IsLoaded=false` until an entry is selected).
- **Open Source File** / **Open Source Folder** correctly hidden (`!hasSource` — no previously imported file recorded).
- All other controls from PR #506 still render: read-config bar, selection bar, list filter, edit fields (D0/W4/W6/Comment), List Expansion (disabled, gated on #501).

### Environment
- ROM: `roms/FE8U.gba` (Fire Emblem - The Sacred Stones, USA — vanilla, no map-action-animation patch installed).
- View: `ImageMapActionAnimationView` (Avalonia).
- Capture method: PrintWindow API (PW_RENDERFULLCONTENT) + PowerShell UIAutomationClient navigation.

### Steps Performed
1. Launched the Avalonia app via `Start-Process FEBuilderGBA.Avalonia.exe -ArgumentList "--rom","test_rom.gba"`.
2. Located the "Map Action Anim" button on the main menu via UIAutomation (`Main_MapActionAnim_Button`) and invoked it.
3. Verified the opened window is `ImageMapActionAnimationView` (title = "Map Action Animation").
4. Resized the window to 1200x850 to ensure full content layout.
5. Captured the editor window via `tools/WinCapture` (PrintWindow API with PW_RENDERFULLCONTENT).
6. Reviewed the saved PNG and confirmed all four new buttons render with the correct disabled / hidden default states.

### Test Results
| Test | Expected | Actual | Status |
|------|----------|--------|--------|
| Main menu opens upgraded view | `ImageMapActionAnimationView` window with title "Map Action Animation" | Window opened, MainWindowTitle confirmed | PASS |
| Export Animation button rendered | Button visible in toolbar, disabled until valid animation selected | Button visible in screenshot, IsEnabled=False | PASS |
| Import Animation button rendered | Button visible in toolbar, disabled until non-empty entry selected | Button visible in screenshot, IsEnabled=False | PASS |
| Open Source File button rendered | Hidden when no source path recorded | Not visible in screenshot (IsVisible=False) | PASS |
| Open Source Folder button rendered | Hidden when no source path recorded | Not visible in screenshot (IsVisible=False) | PASS |
| All PR #506 controls preserved | Read-config bar, selection bar, edit fields, list expansion still present | All controls visible in screenshot | PASS |
| `L10nCoverageTest` | 0 untranslated literals after the 6 new strings (4 buttons + 2 tooltips) translated to ja+zh | Passes | PASS |
| `ImageMapActionAnimationParityTests` | 23 cases including 8 new #499-specific tests | 23 pass, 0 fail | PASS |
| `MapActionAnimationExportImportCoreTests` | 10 cases covering Export/Import API surface + edge cases | 10 pass, 0 fail | PASS |
| `GifEncoderCoreTests` | 9 cases covering rounding helper + encode + IndexedToRgba | 9 pass, 0 fail | PASS |

### Verdict
**PASS** — Every new control surface renders. The disabled / hidden default states correctly reflect the empty-list state on a vanilla FE8U ROM (matching the WinForms equivalent on the same input). All automated assertions pass.

## Test plan

- [x] `dotnet test FEBuilderGBA.Avalonia.Tests/` — **6355 pass, 3 skipped, 0 fail.** Includes 8 new `ImageMapActionAnimationParityTests` cases for #499.
- [x] `dotnet test FEBuilderGBA.Core.Tests/` — **1758 pass, 0 fail.** Includes 10 new `MapActionAnimationExportImportCoreTests` + 9 new `GifEncoderCoreTests` cases.
- [x] `dotnet build FEBuilderGBA.Core/FEBuilderGBA.Core.csproj` — succeeded.
- [x] `dotnet build FEBuilderGBA.CLI/FEBuilderGBA.CLI.csproj` — succeeded (uses new `GifEncoderCore`).
- [x] `dotnet build FEBuilderGBA.SkiaSharp/FEBuilderGBA.SkiaSharp.csproj` — succeeded.
- [x] `dotnet build FEBuilderGBA.Avalonia/FEBuilderGBA.Avalonia.csproj -c Release` — succeeded.
- [x] `L10nCoverageTest` passes — all 4 new button literals + 2 tooltip literals translated in ja + zh (`Export Animation` -> `アニメーション取出` / `导出动画`; `Import Animation` -> `アニメーション読込` / `导入动画`; tooltips + Open/Select Source preserved from #506).
- [x] GUI screenshot captured via `PrintWindow` against the live Avalonia app with `roms/FE8U.gba` (see Screenshots section). Committed to master via #618 so the URL stays permanent after this branch is deleted.
- [x] Copilot CLI plan review v3 accepted — no remaining blockers (passes 1 + 2 + 3 all green).

## Known limitations

- **Vanilla FE8U does not install the map-action-animation patch.** The Export/Import buttons are correctly rendered in their disabled state on this ROM. To functionally test Export/Import end-to-end requires a ROM with the patch installed; the Core round-trip test (`ImportScript_RawPath_PreservesFrameTablePointers`) exercises the equivalent code path on a synthetic ROM. Headless tests cover all error / RAW / quantize-fallback / palette-padding paths.
- **`Recycle*Anime` from WF Import not implemented in Core path.** The Avalonia Import uses `FindAndWriteData` for free-space placement; orphaned old sheets/palettes stay in ROM. Reclamation is tracked by the existing `MakeAllDataLength` pass at full-ROM rebuild time (same caveat as the WF path's recycle dependencies).
- **Drag-drop of `.txt` files onto the editor not implemented.** WF has `U.AllowDropFilename`; Avalonia has its own drag-drop infrastructure that's wired separately. Out of scope here.
- **`ToolAnimationCreator` "jump to editor" wiring (#500) and List Expansion (#501) remain open** — neither blocks #499.

Generated with Claude Code (claude-opus-4-7[1m])

### CODE REVIEW

File: FEBuilderGBA.Core/MapActionAnimationExportImportCore.cs [hunk 1]
```diff
@@ -0,0 +1,588 @@
...
+                string pngName = sp[1];
+                uint sound = sp.Length >= 3 ? U.atoi0x(sp[2]) : 0;
+
+                byte[] tileBytes;
+                byte[] paletteBytes;
+
+                if (pendingRawTiles != null && pendingRawPalette != null)
+                {
+                    tileBytes = HexToBytes(pendingRawTiles);
+                    paletteBytes = PadGBAPaletteTo16(HexToBytes(pendingRawPalette));
```
Comment (laqieer) [inline FEBuilderGBA.Core/MapActionAnimationExportImportCore.cs]: `HexToBytes(pendingRawPalette)` can return `null` for malformed RAW-PALETTE hex, but passing that directly into `PadGBAPaletteTo16` silently converts the invalid palette into 32 zero bytes. That means a typo in an exported/imported script mutates the ROM and produces black/transparent frames instead of failing like the RAW-TILES path below. Please parse the palette into a temporary, reject `null`, then pad only valid palette bytes, and add a regression test for malformed RAW-PALETTE.

**Review by laqieer — COMMENTED** [review]
Blocking issues found (I cannot submit this as a formal “request changes” review because GitHub disallows requesting changes on your own PR, so I am submitting it as a review comment with blocking findings):

1. The screenshot evidence does not satisfy the GUI feat/fix policy. The PR modifies `FEBuilderGBA.Avalonia/` and the title starts with `feat`, but the provided image shows an empty `ImageMapActionAnimationView` (`0 items`, blank fields) with Export/Import disabled. It is a real app screenshot and the URL is a valid default-branch raw URL, but it does not show populated data or demonstrate the export/import behavior this PR adds. Please add/replace with a relevant running-app screenshot using populated map action animation data, showing the changed editor state/behavior rather than only an empty disabled default state.

2. Inline comment: malformed `# RAW-PALETTE:` hex is silently accepted as a zero-filled palette, which can corrupt an import instead of failing fast. Please validate it like `RAW-TILES` and add a regression test.

I verified the PR body has a `## GUI Test Report` table and the `## Test plan` items are all checked; those parts are not blocked.

Copilot CLI: 1.0.54
Model: GPT-5.5 (gpt-5.5)

**Review by laqieer — COMMENTED** [review]
Re-review after commit `45a864970887a3fd65814a888815defcf583d056`: no remaining blocking issues from my previous review.

Verified the five prior inline blockers are addressed:

1. GIF export is now reachable from the normal SaveFile picker via a dedicated `*.gif` file type choice.
2. The Export/Import button-state documentation now matches the Avalonia behavior: buttons remain rendered and are enabled/disabled rather than hidden.
3. Malformed `# RAW-PALETTE:` hex now returns an explicit import error instead of silently padding `null` into an all-zero palette, with regression coverage.
4. `pointerAddr` is validated as an in-bounds 4-byte pointer slot before `rom.write_p32`, with regression coverage.
5. Imported `Wait` and `Sound` values are range-checked before serialization to u8/u16, with regression coverage for over-range values.

Also verified the screenshot blocker is resolved. PR #624 is merged, `master` contains `pr-screenshots/pr499-mapactionanim-export-import.png` at blob SHA `7876d21f6d88362071b56dacfd5fef4860c1cbef`, and the raw URL now resolves to the updated 1200x850 image showing 5 populated entries, selected `0x01 Break1`, rendered preview data, and enabled Export Animation / Import Animation buttons.

CI is green on the latest PR head. The only remaining open review thread I see is an outdated duplicate self-authored RAW-PALETTE thread; the active Copilot bot threads are resolved and the code now covers that case.

Copilot CLI: 1.0.54
Model: GPT-5.5 (gpt-5.5)
